### PR TITLE
Checkmarx Scan Failure Should Create a JIRA Ticket

### PR DESCRIPTION
**Pull request by dealako**
* Added logic to check the build result and if unsuccessful throws a
REException which is caught in the calling function. From there the
common code will create a JIRA ticket if the JIRA_PROJECT_KEY is set.

JIRA: RE-2388

Issue: [RE-2388](https://rpc-openstack.atlassian.net/browse/RE-2388)

### CODE REVIEW

File: pipeline_steps/checkmarx.groovy [hunk 1]
```diff
@@ -80,6 +80,10 @@ def scan(String scan_type, String repo_name, String exclude_folders){
                 // exception must propagate back to the retry call
                 throw e
             } //try
+
+            if (currentBuild.getResult() && currentBuild.getResult().isWorseThan(SUCCESS)) {
```
Comment (git-harry) [inline pipeline_steps/checkmarx.groovy]: The [documentation](https://rpc.jenkins.cit.rackspace.net/job/PM-Checkmarx_all-rpc-differ/pipeline-syntax/globals#currentBuild) doesn't list `currentBuild` as supporting this method but has others that are similar, has this been tested? This looks like something probably found in `currentBuild.rawBuild`.

File: pipeline_steps/checkmarx.groovy [hunk 2]
```diff
@@ -80,6 +80,10 @@ def scan(String scan_type, String repo_name, String exclude_folders){
                 // exception must propagate back to the retry call
                 throw e
             } //try
+
```
Comment (git-harry) [inline pipeline_steps/checkmarx.groovy]: This following block is within the retry block so it's going to cause unintentional retries when there is a genuine failure.

File: pipeline_steps/checkmarx.groovy [hunk 3]
```diff
@@ -80,6 +80,10 @@ def scan(String scan_type, String repo_name, String exclude_folders){
...
+            if (currentBuild.getResult() && currentBuild.getResult().isWorseThan(SUCCESS)) {
+                throw new REException("Checkmarx Scan Threshold Exceeded for repo: " + repo_name)
```
Comment (git-harry) [inline pipeline_steps/checkmarx.groovy]: Does REException need importing?

File: pipeline_steps/checkmarx.groovy (hunk 2 above)
Comment (dealako) [reply pipeline_steps/checkmarx.groovy]: Yes, will move below the retry logic ✅

File: pipeline_steps/checkmarx.groovy (hunk 3 above)
Comment (dealako) [reply pipeline_steps/checkmarx.groovy]: Yes

File: pipeline_steps/checkmarx.groovy (hunk 1 above)
Comment (dealako) [reply pipeline_steps/checkmarx.groovy]: The documentation link is helpful. Pushed update to hopefully address the usage of this object.

File: pipeline_steps/checkmarx.groovy [hunk 4]
```diff
@@ -81,6 +82,10 @@ def scan(String scan_type, String repo_name, String exclude_folders){
                 throw e
             } //try
         } // retry
+
+        if (currentBuild.getRawBuild().getResult().isWorseThan(Result.fromString("SUCCESS"))) {
```
Comment (git-harry) [inline pipeline_steps/checkmarx.groovy]: If the scan is successful `currentBuild.getRawBuild().getResult() == null` which causes this line to raise an exception. I would suggest:
```suggestion
        if (Result.fromString(currentBuild.currentResult).isWorseThan(Result.fromString("SUCCESS"))) {
```

**Pull request merged**